### DISCUSSION TITLE
Feature Request: "Fugue Mode" -> multiple playheads / step time division / playhead speed / playback direction

### DISCUSSION THREAD

**Discussion by andreacfromtheapp** [SynthstromAudible/DelugeFirmware#1619] [Ideas]
Hi :)

first things first: thank you for the amazing work you do.

secondly: it would be awesome if Deluge had a "fugue mode" like the iOS app [Fugue Machine](https://alexandernaut.com/fuguemachine/).

people seem to want this bad:

- https://forums.synthstrom.com/discussion/5261/multiple-playheads-or-tempo-math-per-clip-fugue-machine/p1
- https://forums.synthstrom.com/discussion/comment/9786#Comment_9786
- https://forums.synthstrom.com/discussion/comment/8857#Comment_8857
- https://forums.synthstrom.com/discussion/879/multiple-playheads-mode/p1
- https://forums.synthstrom.com/discussion/2240/100-top-rated-suggestions-2017-2019/p1

Please and thank you :)

**Comment by joxihan** [comment]
There is the same feature on Circuit Tracks and I confirm it's a lot of fun !

**Comment by andreacfromtheapp** [comment]
Hi,

this is being discussed on Deluge Discord -> #feature-forum -> Fugue Mode (Multi-playhead Sequencer with independent subdivisions).

I have also prepared the following "spec" to better illustrate my idea of how it could work on the Deluge. Pasting this here for better readibility and because Discord character limits. I have also attached the markdown file on Discord for anyone to download.

# Fugue mode feature



- [Fugue mode feature](#fugue-mode-feature)
  - [Terminology](#terminology)
    - [Fugue mode](#fugue-mode)
    - [Fugue view](#fugue-view)
    - [Playhead](#playhead)
    - [Fugue clip](#fugue-clip)
  - [Community features settings](#community-features-settings)
    - [Limitations](#limitations)
    - [Proposed settings](#proposed-settings)
  - [Affected clip types](#affected-clip-types)
  - [Affected existing features](#affected-existing-features)
    - [DIRECTION](#direction)
    - [CLIPS, DISPLAY, KNOBS, COMBOS, UI/UX](#clips-display-knobs-combos-uiux)
  - [Playhead properties](#playhead-properties)
    - [Common playhead parameters](#common-playhead-parameters)
    - [Independent playhead parameters](#independent-playhead-parameters)
  - [Workflow](#workflow)
    - [How to create a fugue clip](#how-to-create-a-fugue-clip)
    - [How to delete a fugue clip](#how-to-delete-a-fugue-clip)
    - [How to navigate fugue clips](#how-to-navigate-fugue-clips)
    - [How to edit playheads](#how-to-edit-playheads)
      - [Affecting ALL PLAYHEADS at once:](#affecting-all-playheads-at-once)
      - [Affecting the currently viewed/edited PLAYHEAD:](#affecting-the-currently-viewededited-playhead)
    - [How to start/stop playheads](#how-to-startstop-playheads)




Notes to showcase the feature and define its UI/UX, as detailed as possible.
With its workflow, UI, user feedback, and _properties_. As well as affected
features, clips, limitations, suggestions, community features settings, and
more.

## Terminology

### Fugue mode

This is the feature idea in and of itself, encompassing all that is hereby
described. Characterized as FUGUE MODE.

### Fugue view

A GRID view similar to TRIPLETS VIEW. All details in _Workflow_. Characterized
as FUGUE VIEW.

### Playhead

I originally referred to this as _clip pages_ and _subclips_, in my Discord
request. Characterized as PLAYHEAD(S).

### Fugue clip

To differentiate a _CLIP_ as it's currently implemented from a _FUGUE VIEW
enabled CLIP_, the former shall be referred as CLIP in current context. In a
FUGUE VIEW context, the former is referred as PARENT CLIP and the latter as
FUGUE CLIP.

## Community features settings

[Fugue Machine](https://alexandernaut.com/fugue-machine/) (the inspiration for
FUGUE MODE) has a fixed number (4) of playheads with its only (1) piano roll.
This would translate into 4 PLAYHEADS per FUGUE CLIP on the Deluge.

### Limitations

Setting some limits to FUGUE MODE may beneficial. Like Fugue Machine, we could
impose 4 PLAYHEADS per FUGUE CLIP and no settings at all. Limiting the number of
available FUGUE CLIPS per project could also be an option to limit resource
usage.

**Alternatively**: leave it to the user to set PLAYHEADS limits per FUGUE CLIP.
Because this may be resource hungry, and potentially sonically messy, default to
4 and limit to 8 per FUGUE CLIP. Also letting users enable/disable affected clip
types and decide how many FUGUE CLIPS per project could be nice.

### Proposed settings

Defaults in **bold**:

- COMMUNITY FEATURES -> FUGUE MODE -> FUGUE CLIP TYPES -> **MIDI [x]**, CV [ ],
  **SYNTH [x]**, AUDIO [ ]
- COMMUNITY FEATURES -> FUGUE MODE -> PLAYHEADS PER CLIP: [**4**, 8]
- COMMUNITY FEATURES -> FUGUE MODE -> MAX FUGUE CLIPS PER PROJECT: [**2**, 4]

## Affected clip types

For all intention and purposes a PLAYHEAD triggers notes normally, regardless of
the play direction. Should the already implemented DIRECTION feature support it,
none of these should be an issue. However, because in a fugue multiple notes may
be triggered at the same time, this could be problematic for some CLIP types.

- MIDI
- CV
- SYNTH (uncertain how this would affect the envelope tho)
- perhaps AUDIO (if bidirectional audio playback is supported)

## Affected existing features

### DIRECTION

The current DIRECTION play modes are:

- Forward
- Reverse
- Ping Pong

Implement these additional play modes (extend/substitute Ping Pong with them):

- Forward-Reverse (Ping Pong?)
- Reverse-Forward (Reverse Ping Pong?)

### CLIPS, DISPLAY, KNOBS, COMBOS, UI/UX

**Note**: the following bird's eye view sounds scarier than it is. The
_Workflow_ section details it all from a practical step by step perspective.

Modifications to CLIPS, DISPLAY, KNOBS, COMBOS, UI and UX, are necessary to
enable the following, to convey their meaning, and to provide visual feedback
and clear definitions to end users:

- to enable FUGUE VIEW for a CLIP
- to navigate FUGUE CLIP/PLAYHEADS
- to edit PLAYHEADS parameters
- to blink/alter related LEDs
- to show related DISPLAY entries
- to configure related Community Features

## Playhead properties

**With some exceptions (see _Common playhead parameters_)**:

- PLAYHEADS behave like a CLIP for all intentions and purposes
- PLAYHEADS are independently editable as if they were a CLIP
- Most existing shortcuts to edit PLAYHEADS work as they do with a CLIPS
- New shortcuts and keys/knobs combos detailed in the _Workflow_ section

### Common playhead parameters

ALL PLAYHEADS must retain these PARENT CLIP parameters at the same time:

- **TYPE**: all PLAYHEADS must be of the same TYPE. If the PARENT CLIP was MIDI
  so are ALL PLAYHEADS. Changing TYPE (or SYNTH patch or MIDI channel and so on)
  for a single PLAYHEAD, should reflect that for ALL PLAYHEADS
- **MEASURES/BARS**: all PLAYHEADS must be of the same LENGTH. If the PARENT
  CLIP was 16 bars, so are ALL PLAYHEADS. Changing LENGTH for a single PLAYHEAD
  should reflect that for ALL PLAYHEADS
- **SCALE**: all PLAYHEADS must be of the same SCALE. Changing SCALE for a
  single PLAYHEAD should reflect that for ALL PLAYHEADS
- **NOTES**: all PLAYHEADS must have the same NOTES. Changing NOTES for a single
  PLAYHEAD should reflect that for ALL PLAYHEADS

**Note**: there may be even more parameters that need to be COMMON. I tried to
think extensively but may have forgotten about or missed something.

### Independent playhead parameters

For convenience, ALL PLAYHEADS start by holding the same values for all of the
following parameters, derived from the PARENT CLIP. However, each PLAYHEAD
allows editing them independently to honor how fugues work. Each PLAYHEAD has
dedicated controls that define how the PLAYHEAD scans the musical sequence.
These controls are:

- **DIRECTION**: play the sequence forward, reverse, forward-reverse, or
  reverse-forward
- **TEMPO**: speed of the playhead relative to master tempo
  - NOTE: default tempo 1x (meaning same as master tempo)
- **OCTAVE**: transpose up and down by octaves
- **SEMI**: transpose up and down by semitones
- **VELOCITY**: loudness range of each playhead
  - NOTE: not sure velocity ranges are feasible
- **START POSITION**: step at which playhead starts playing the sequence
  - NOTE: defaults at step 1.1.1
- **INVERT SWITCH**: if enabled, the sequence will be harmonically inverted. In
  other words, the highest pitch becomes the lowest pitch, and the lowest pitch
  becomes the highest pitch
  - NOTE: **if this is too much work, leave it out**

## Workflow

I'll be using **PLAYHEADS PER CLIP = 4** throughout this Workflow section and in
all PNG mockups. See _Proposed settings_.

### How to create a fugue clip

1. select CLIP
2. select CLIP TYPE (i.e: MIDI)
3. set CLIP length (MEASURES/BARS)
4. enter NOTES
5. press SHIFT + TRIPLETS VIEW (the **CLIP** becomes a **FUGUE CLIP**)
   - **UI -> LED:** TRIPLETS VIEW LED BLINKS
   - **UI -> GRID:** akin to TRIPLETS VIEW, FUGUE VIEW should be represented on
     the GRID [1]
   - **CLIP -> FUGUE CLIP:** the PARENT CLIP _is deeply copied into_ N
     PLAYHEADS, to form a FUGUE CLIP [2]

[1] **FUGUE VIEW**: a GRID view similar to TRIPLETS VIEW except the following

- the columns are placed at the start of each PLAYHEAD
- each column should have a different color to differentiate PLAYHEADS

<img width="800" alt="FUGUE-VIEW" src="https://github.com/user-attachments/assets/5256a256-e314-466d-bb51-64f781a01e9f" />

[2] **FUGUE CLIP**: a PARENT CLIP that has been _multiplied by PLAYHEADS PER
CLIP_ . For example:

- with a setting of PLAYHEADS PER CLIP = 4, a 16 bars PARENT CLIP becomes a 64
  bars FUGUE CLIP
- all PLAYHEADS start with the same properties and values as the PARENT CLIP
  (see _Playhead properties_)

In other words, the resulting FUGUE CLIP comprises a 1:1 copy of the PARENT CLIP
for each of the 4 PLAYHEADS

**IMPORTANT:** the PARENT CLIP is irreversibly lost upon conversion to FUGUE
CLIP.

### How to delete a fugue clip

- **Delete a FUGUE CLIP:** from GRID/ROWS, deleting a FUGUE CLIP works just like
  a normal CLIP.

**IMPORTANT:** Because converting a FUGUE CLIP back to a CLIP would be
unfeasible - (Which PLAYHEAD to keep/discard? What PLAYHEAD parameters to
keep/discard? And so on) - FUGUE CLIPS should remains so until deleted from the
GRID/ROWS.

### How to navigate fugue clips

- **Identify FUGUE CLIPS:** from GRID/ROWS, press and hold TRIPLETS VIEW. All
  FUGUE CLIPS should blink until TRIPLETS VIEW is released
- **Enter a FUGUE CLIP:** from GRID/ROWS, enter a FUGUE CLIP just like you would
  a normal CLIP.
  - **UI -> LED:** TRIPLETS VIEW LED BLINKS
  - **UI -> GRID:** FUGUE VIEW is displayed
- **Enter a PLAYHEAD:** from FUGUE CLIP, press any PAD within the PLAYHEAD range
- **Exit a PLAYHEAD:** (aka Go Back to FUGUE CLIP): from PLAYHEAD, press LEARN +
  TRIPLETS VIEW
- **Exit a FUGUE CLIP:** (aka Go Back to GRID/ROWS): from FUGUE CLIP or
  PLAYHEAD, exit just like you would a normal CLIP

### How to edit playheads

**IMPORTANT:** FUGUE CLIPS are NOT directly editable. Instead, edit PLAYHEADS.

Editing a PLAYHEAD works just like with a normal CLIP. However, some parameters
will affect ALL PLAYHEADS, and some other will work for each PLAYHEAD
independently (see _Playhaed properties_). Hereby the shortctus, knobs, and
combos (old and new).

#### Affecting ALL PLAYHEADS at once:

- **TYPE**: select SYNTH, MIDI, CV, AUDIO as usual (see _Community features
  settings_)
- **MEASURES/BARS**: edit length as usual
- **SCALE**: select scale and root note as usual
- **NOTES**: edit notes as usual

#### Affecting the currently viewed/edited PLAYHEAD:

- **DIRECTION**: set/change the play direction as usual
- **OCTAVE**: set/change the octave as usual
- **SEMI**: set/change the semitone as usual
- **VELOCITY**: set/change velocity as usual
- **TEMPO**: press and hold TRIPLETS VIEW + rotate TEMPO knob
  - **UI -> DISPLAY:** display available tempo options (2x, 1x, 1/2, 1/4, 1/8,
    etc...)
- **START POSITION**: press and hold TRIPLETS VIEW + press desired PAD
  corresponding to the bar
  - NOTE: PLAYHEAD should be zoomed out/users need to navigate to desired bar
  - **UI -> GRID:** selected START POSITION highlighted with a white column
- **INVERT SWITCH**: **I woulnd't know how to do this nor if it's possible at
  all** It's anybody's guess.

### How to start/stop playheads

- **Start/Stop a PLAYHEAD:** from FUGUE CLIP, press and hold LEARN + press any
  PAD within the PLAYHEAD range
  - **UI -> FUGUE VIEW:** playing PLAYHEADS BLINKING (invert for stopped
    PLAYHEADS)
